cocos/core/pipeline/pipeline-state-manager.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @hidden
+ */
+
+import { GFXPipelineState, IGFXPipelineStateInfo } from '../gfx/pipeline-state';
+import { IPSOCreationInfo } from '../renderer/scene/submodel';
+import { GFXRenderPass } from '../gfx/render-pass';
+import { GFXInputAssembler } from '../gfx/input-assembler';
+import { murmurhash2_32_gc } from '../utils/murmurhash2_gc';
+import { GFXDevice } from '../gfx/device';
+
+export class PipelineStateManager {
```

Comment:
Do we need to detect useless PSOs and destroy them if no longer needed ? Like a garbage collect mechanism

Reply:
We don't know which PSO is not needed, it may be used later. Of course, we can add timer to record how long a PSO is not used, but it may make things complex. We can destroy PSOs when memory is low if we have the notification. But create PSO is slow, we should cache them as possible. As i know, bgfx doesn't destroy it.

Reply:
I see